README.md
# Spiderable middleware

Google, Facebook, Twitter, Yahoo, and Bing and all other crawlers and search engines are constantly trying to view your website. If your website is built on top of the JavaScript framework like, but not limited to - Angular, Backbone, Ember, Meteor, React, MEAN most of the front-end solutions returns basic HTML-markup and script-tags to crawlers, but not content of your page. The mission of `spiderable-middleware` and [ostr.io](https://ostr.io) are to boost your SEO experience without a headache.

## Why Pre-render?

- 🕸 Execute JavaScript, as web-crawlers and search engines can't run JS code;
- 🏃‍♂️ Boost response rate and decrease response time with caching;
- 🚀 Optimized HTML markup for best SEO score;
- 🖥 Support for PWAs and SPAs;
- 📱 Support for mobile-like crawlers;
- 💅 Support [`styled-components`](https://styled-components.com);
- ⚡️ Support [AMP (Accelerated Mobile Pages)](https://www.ampproject.org);
- 🤓 Works with `Content-Security-Policy` and other "complicated" front-end security;
- ❤️ Search engines and social network crawlers love straightforward and pre-rendered pages;
- 📱 Consistent link previews in messaging apps, like iMessage, Messages, Facebook, Slack, Telegram, WhatsApp, Viber, VK, Twitter, etc.;
- 💻 Image, title, and description previews for posted links at social networks, like Facebook, Twitter, VK and others.

## About Package

This package acts as middleware and intercepts requests to your Node.js application from web crawlers. All requests proxy passed to the Prerendering Service, which returns static, rendered HTML.

__This is SERVER only package. For NPM make sure you're importing library only in Node.js. For Meteor.js please make sure library imported and executed only on SERVER.__

We made this package with developers in mind. It's well written in a very simple way, hackable, and easily tunable to meet your needs, can be used to turn dynamic pages into rendered, cached, and lightweight static pages, just set `botsUA` to `['.*']`. This is the best option to offload servers unless a website gets updated more than once in 4 hours.

- __Note__: *This package proxies real HTTP headers and response code, to reduce overwhelming requests, try to avoid HTTP-redirect headers, like* `Location` *and others. Read how to [return genuine status code](https://github.com/veliovgroup/spiderable-middleware#return-genuine-status-code) and [handle JS-redirects](https://github.com/veliovgroup/spiderable-middleware#javascript-redirects).*
- __Note__: This is __server only package__. For isomorphic environments, *like Meteor.js*, this package should be imported/initialized only at server code base.

This middleware was tested and works like a charm with:

- [meteor](https://www.meteor.com/): [example](https://github.com/veliovgroup/spiderable-middleware/blob/master/examples/meteor.middleware.js)
- [express](https://www.npmjs.com/package/express): [example](https://github.com/veliovgroup/spiderable-middleware/blob/master/examples/express.middleware.js)
- [connect](https://www.npmjs.com/package/connect): [example](https://github.com/veliovgroup/spiderable-middleware/blob/master/examples/connect.middleware.js)
- [vanilla http(s) server](https://nodejs.org/api/http.html): [example](https://github.com/veliovgroup/spiderable-middleware/blob/master/examples/http.middleware.js)
- See [all examples](https://github.com/veliovgroup/spiderable-middleware/tree/master/examples)

All other frameworks which follow Node's middleware convention - will work too.

This package was originally developed for [ostr.io](https://ostr.io) service. But it's not limited to, and can proxy-pass requests to any other rendering-endpoint.

## ToC

- [Installation](https://github.com/veliovgroup/spiderable-middleware#installation)
- [Basic usage](https://github.com/veliovgroup/spiderable-middleware#basic-usage)
- [MeteorJS usage](https://github.com/veliovgroup/spiderable-middleware#meteor-specific-usage)
- [Return genuine status code](https://github.com/veliovgroup/spiderable-middleware#return-genuine-status-code)
- [Speed-up rendering](https://github.com/veliovgroup/spiderable-middleware#speed-up-rendering)
- [Detect request from Prerendering engine during runtime](https://github.com/veliovgroup/spiderable-middleware#detect-request-from-pre-rendering-engine-during-runtime)
- [Detect request from Prerendering engine during __Meteor/Blaze__ runtime](https://github.com/veliovgroup/spiderable-middleware#detect-request-from-pre-rendering-engine-in-meteorjs)
- [JavaScript redirects](https://github.com/veliovgroup/spiderable-middleware#javascript-redirects)
- [AMP Support](https://github.com/veliovgroup/spiderable-middleware#amp-support)
- [Rendering Endpoints](https://github.com/veliovgroup/spiderable-middleware#rendering-endpoints)
- [API](https://github.com/veliovgroup/spiderable-middleware#api)
  - [Constructor](https://github.com/veliovgroup/spiderable-middleware#constructor-new-spiderableopts)
  - [Middleware](https://github.com/veliovgroup/spiderable-middleware#spiderablehandlerreq-res-next)
- [Debugging](https://github.com/veliovgroup/spiderable-middleware#debugging)
- [Running Tests](https://github.com/veliovgroup/spiderable-middleware#running-tests)
  - [Node.js/Mocha](https://github.com/veliovgroup/spiderable-middleware#nodejsmocha)
  - [Meteor/Tinytest](https://github.com/veliovgroup/spiderable-middleware#meteortinytest)

## Installation

This package is distributed via NPM for Node.js and Atmosphere for Meteor.js. Although it is safe to use NPM distributed version in Meteor backend.

### NPM:

```sh
npm install spiderable-middleware --save
```

### Meteor:

```sh
meteor add webapp
meteor add ostrio:spiderable-middleware
```

## Usage

Get ready in a few lines of code

### Basic usage

See [all examples](https://github.com/veliovgroup/spiderable-middleware/tree/master/examples).

First, add `fragment` meta-tag to your HTML template:

```html
<html>
  <head>
    <meta name="fragment" content="!">
    <!-- ... -->
  </head>
  <body>
    <!-- ... -->
  </body>
</html>
```

```js
// Make sure this code isn't exported to the Browser bundle
// and executed only on SERVER (Node.js)
const express    = require('express');
const app        = express();
const Spiderable = require('spiderable-middleware');
const spiderable = new Spiderable({
  rootURL: 'http://example.com',
  serviceURL: 'https://render.ostr.io',
  auth: 'APIUser:APIPass'
});

app.use(spiderable.handler).get('/', (req, res) => {
  res.send('Hello World');
});

app.listen(3000);
```

We provide various options for `serviceURL` as "[Rendering Endpoints](https://github.com/veliovgroup/ostrio/blob/master/docs/prerendering/rendering-endpoints.md)", each endpoint has its own features to fit every project needs.

### Meteor specific usage

```js
// Install necessary packages:
// meteor add webapp
// meteor add ostrio:spiderable-middleware

// Make sure this code executed only on SERVER
// Use `if (Meteor.isServer) {/*...*/}` blocks
// or place this code under `/server/` directory
import { WebApp } from 'meteor/webapp';
import Spiderable from 'meteor/ostrio:spiderable-middleware';

WebApp.connectHandlers.use(new Spiderable({
  rootURL: 'http://example.com',
  serviceURL: 'https://render.ostr.io',
  auth: 'APIUser:APIPass'
}));
```

## Return genuine status code

To pass expected response code from front-end JavaScript framework to browser/crawlers, you need to create specially formatted HTML-comment. This comment __can be placed in any part of HTML-page__. `head` or `body` tag is the best place for it.

### Format

__html__:

```html
<!-- response:status-code=404 -->
```

__jade__:

```jade
// response:status-code=404
```

This package support __any__ standard and custom status codes:

- `201` - `<!-- response:status-code=201 -->`
- `401` - `<!-- response:status-code=401 -->`
- `403` - `<!-- response:status-code=403 -->`
- `500` - `<!-- response:status-code=500 -->`
- `514` - `<!-- response:status-code=514 -->` (*non-standard*)

__Note__: *Reserved status codes for internal service communications:* `49[0-9]`.

## Speed-up rendering

To speed-up rendering, you __should__ tell to the Spiderable engine when your page is ready. Set `window.IS_RENDERED` to `false`, and once your page is ready set this variable to `true`. Example:

```html
<html>
  <head>
    <meta name="fragment" content="!">
    <script>
      window.IS_RENDERED = false;
    </script>
  </head>
  <body>
    <!-- ... -->
    <script type="text/javascript">
      //Somewhere deep in your app-code:
      window.IS_RENDERED = true;
    </script>
  </body>
</html>
```

## Detect request from Pre-rendering engine during runtime

Pre-rendering engine will set `window.IS_PRERENDERING` global variable to `true`. Detecting requests from pre-rendering engine are as easy as:

```js
if (window.IS_PRERENDERING) {
  // This request is coming from Pre-rendering engine
}
```

__Note__: `window.IS_PRERENDERING` can be `undefined` on initial page load, and may change during runtime. That's why we recommend to pre-define a setter for `IS_PRERENDERING`:

```js
let isPrerendering = false;
Object.defineProperty(window, 'IS_PRERENDERING', {
  set(val) {
    isPrerendering = val;
    if (isPrerendering === true) {
      // This request is coming from Pre-rendering engine
    }
  },
  get() {
    return isPrerendering;
  }
});
```

## Detect request from Pre-rendering engine in Meteor.js

Pre-rendering engine will set `window.IS_PRERENDERING` global variable to `true`. As in Meteor/Blaze everything should be reactive, let's bound it with `ReactiveVar`:

```js
import { Template }    from 'meteor/templating';
import { ReactiveVar } from 'meteor/reactive-var';

const isPrerendering = new ReactiveVar(window.IS_PRERENDERING || false);
Object.defineProperty(window, 'IS_PRERENDERING', {
  set(val) {
    isPrerendering.set(val);
  },
  get() {
    return isPrerendering.get();
  }
});

// Make globally available Blaze helper,
// Feel free to omit this line in case if you're not using Blaze
// or going to handle logic in JavaScript part
Template.registerHelper('IS_PRERENDERING', () => isPrerendering.get());
```

__Note__: `window.IS_PRERENDERING` can be `undefined` on initial page load, and may change during runtime.

## JavaScript redirects

If you need to redirect browser/crawler inside your application, while a page is loading (*imitate navigation*), you're free to use any of classic JS-redirects as well as your framework's navigation, or `History.pushState()`

```js
window.location.href = 'http://example.com/another/page';
window.location.replace('http://example.com/another/page');

Router.go('/another/page'); // framework's navigation !pseudo code
```

__Note__: *Only 4 redirects are allowed during one request after 4 redirects session will be terminated.*

## API

### *Constructor* `new Spiderable([opts])`

- `opts` {*Object*} - Configuration options
- `opts.serviceURL` {*String*} - Valid URL to Spiderable endpoint (local or foreign). Default: `https://render.ostr.io`. Can be set via environment variables: `SPIDERABLE_SERVICE_URL` or `PRERENDER_SERVICE_URL`
- `opts.rootURL` {*String*} - Valid root URL of your website. Can be set via an environment variable: `ROOT_URL` (*common for meteor*)
- `opts.auth` {*String*} - [Optional] Auth string in next format: `user:pass`. Can be set via an environment variables: `SPIDERABLE_SERVICE_AUTH` or `PRERENDER_SERVICE_AUTH`. Default `null`
- `opts.botsUA` {*[String]*} - [Optional] An array of strings (case insensitive) with additional User-Agent names of crawlers you would like to intercept. See default [bot's names](https://github.com/veliovgroup/spiderable-middleware/blob/master/lib/index.js#L119). Set to `['.*']` to match all browsers and robots, to serve static pages to all users/visitors
- `opts.ignoredHeaders` {*[String]*} - [Optional] An array of strings (case insensitive) with HTTP header names to exclude from response. See default [list of ignored headers](https://github.com/veliovgroup/spiderable-middleware/blob/master/lib/index.js#L121). Set to `['.*']` to ignore all headers
- `opts.ignore` {*[String]*} - [Optional] An array of strings (case __sensitive__) with ignored routes. Note: it's based on first match, so route `/users` will cause ignoring of `/part/users/part`, `/users/_id` and `/list/of/users`, but not `/user/_id` or `/list/of/blocked-users`. Default `null`
- `opts.only` {*[String|RegExp]*} - [Optional] An array of strings (case __sensitive__) or regular expressions (*could be mixed*). Define __exclusive__ route rules for pre-rendering. Could be used with `opts.onlyRE` rules. __Note:__ To define "safe" rules as {*RegExp*} it should start with `^` and end with `$` symbols, examples: `[/^\/articles\/?$/, /^\/article\/[A-z0-9]{16}\/?$/]`
- `opts.onlyRE` {*RegExp*} - [Optional] Regular Expression with __exclusive__ route rules for pre-rendering. Could be used with `opts.only` rules
- `opts.timeout` {*Number*} - [Optional] Number, proxy-request timeout to rendering endpoint in milliseconds. Default: `180000`
- `opts.requestOptions` {*Object*} - [Optional] Options for request module (like: `timeout`, `debug`, `proxy`), for all available options see [`request-libcurl` API docs](https://github.com/veliovgroup/request-extra#request-options)
- `opts.debug` {*Boolean*} - [Optional] Enable debug and extra logging, default: `false`

__Note:__ *Setting* `.onlyRE` *and/or* `.only` *rules are highly recommended. Otherwise, all routes, including randomly generated by bots will be subject of Pre-rendering and may cause unexpectedly higher usage.*

```js
// CommonJS
// const Spiderable = require('spiderable-middleware');

// ES6 import
// import Spiderable from 'spiderable-middleware';

// ES6 import (Meteor.js)
// import Spiderable from 'meteor/ostrio:spiderable-middleware';

const spiderable = new Spiderable({
  rootURL: 'http://example.com',
  serviceURL: 'https://render.ostr.io',
  auth: 'APIUser:APIPass'
});

// More complex setup (recommended):
const spiderable = new Spiderable({
  rootURL: 'http://example.com',
  serviceURL: 'https://render.ostr.io',
  auth: 'APIUser:APIPass',
  only: [
    /\/?/, // Root of the website
    /^\/posts\/?$/, // "/posts" path with(out) trailing slash
    /^\/post\/[A-z0-9]{16}\/?$/ // "/post/:id" path with(out) trailing slash
  ],
  // [Optionally] force ignore for secret paths:
  ignore: [
    '/account/', // Ignore all routes under "/account/*" path
    '/billing/' // Ignore all routes under "/billing/*" path
  ]
});
```

### `spiderable.handler(req, res, next)`

*Middleware handler. Alias:* `spiderable.handle`.

```js
// Express, Connect:
app.use(spiderable.handler);

// Meteor:
WebApp.connectHandlers.use(spiderable);

//HTTP(s) Server
http.createServer((req, res) => {
  spiderable.handler(req, res, () => {
    // Callback, triggered if this request
    // is not a subject of spiderable pre-rendering
    res.writeHead(200, {'Content-Type': 'text/plain; charset=UTF-8'});
    res.end('Hello vanilla NodeJS!');
    // Or do something else ...
  });
}).listen(3000);
```

## AMP Support

To properly serve pages for [Accelerated Mobile Pages](https://www.ampproject.org) (AMP) we support following URI schemes:

```sh
# Regular URIs:
https://example.com/index.html
https://example.com/articles/article-title.html
https://example.com/articles/article-uniq-id/article-slug

# AMP optimized URIs (prefix):
https://example.com/amp/index.html
https://example.com/amp/articles/article-title.html
https://example.com/amp/articles/article-uniq-id/article-slug

# AMP optimized URIs (extension):
https://example.com/amp/index.amp.html
https://example.com/amp/articles/article-title.amp.html
```

All URLs with `.amp.` extension and `/amp/` prefix will be optimized for AMP.

## Rendering Endpoints

- __render (*default*)__ - `https://render.ostr.io` - This endpoint has "optimal" settings, and should fit 98% cases. Respects cache headers of both Crawler and your server;
- __render-bypass (*devel/debug*)__ - `https://render-bypass.ostr.io` - This endpoint has bypass caching mechanism (*almost no cache*). Use it if you're experiencing an issue, or during development, to make sure you're not running into the intermediate cache. You're safe to use this endpoint in production, but it may result in higher usage and response time.
- __render-cache (*under attack*)__ - `https://render-cache.ostr.io` - This endpoint has the most aggressive caching mechanism. Use it if you're looking for the shortest response time, and don't really care about outdated pages in cache for 6-12 hours

To change default endpoint, grab [integration examples code](https://github.com/veliovgroup/spiderable-middleware/tree/master/examples) and replace `render.ostr.io`, with endpoint of your choice. For NPM/Meteor integration change value of [`serviceURL`](https://github.com/veliovgroup/spiderable-middleware#basic-usage) option.


HISTORY.md
For full package history see [releases on GitHub](https://github.com/veliovgroup/spiderable-middleware/releases)


lib/index.js
var url = require('url');
var request = require('request-libcurl');

if (typeof window !== 'undefined') {
  throw new Error('Running `spiderable-middleware` in Browser environment isn\'t allowed! Please make sure `spiderable-middleware` NPM package is imported and used only in Node.js environment.');
}

var strs = {
  pipe: '|',
  empty: '',
  semicolon: ':',
  head: 'head',
  get: 'get',
  ua: 'user-agent',
  caseInsensitive: 'i',
  objs: {
    array: '[object Array]',
    regexp: '[object RegExp]',
    object: '[object Object]',
    string: '[object String]'
  }
};

var re = {
  newLine: /\r|\n/,
  proto: /^https?:\/\//i,
  trailingSlash: /\/$/,
  beginningSlash: /^\//,
  digit: /\d+/,
  staticExt: /\.(?:3ds|3g2|3gp|3gpp|7z|a|aac|aaf|adp|ai|aif|aiff|alz|ape|apk|appcache|ar|arj|asf|asx|atom|au|avchd|avi|bak|bbaw|bh|bin|bk|bmp|btif|bz2|bzip2|cab|caf|cco|cgm|class|cmx|cpio|cr2|crt|crx|css|csv|cur|dat|deb|der|dex|djvu|dll|dmg|dng|doc|docm|docx|dot|dotm|dra|drc|DS_Store|dsk|dts|dtshd|dvb|dwg|dxf|ear|ecelp4800|ecelp7470|ecelp9600|egg|eol|eot|eps|epub|exe|f4a|f4b|f4p|f4v|fbs|fh|fla|flac|fli|flv|fpx|fst|fvt|g3|geojson|gif|graffle|gz|gzip|h261|h263|h264|hqx|htc|ico|ief|img|ipa|iso|jad|jar|jardiff|jng|jnlp|jpeg|jpg|jpgv|jpm|js|jxr|key|kml|kmz|ktx|less|lha|lvp|lz|lzh|lzma|lzo|m2v|m3u|m4a|m4p|m4v|map|manifest|mar|markdown|md|mdi|mdown|mdwn|mht|mid|midi|mj2|mka|mkd|mkdn|mkdown|mkv|mml|mmr|mng|mobi|mov|movie|mp2|mp3|mp4|mp4a|mpe|mpeg|mpg|mpga|mpv|msi|msm|msp|mxf|mxu|nef|npx|nsv|numbers|o|oex|oga|ogg|ogv|opus|otf|pages|pbm|pcx|pdb|pdf|pea|pem|pgm|pic|pl|pm|png|pnm|pot|potm|potx|ppa|ppam|ppm|pps|ppsm|ppsx|ppt|pptm|pptx|prc|ps|psd|pya|pyc|pyo|pyv|qt|ra|rar|ras|raw|rdf|rgb|rip|rlc|rm|rmf|rmvb|ron|roq|rpm|rss|rtf|run|rz|s3m|s7z|safariextz|scpt|sea|sgi|shar|sil|sit|slk|smv|so|sub|svg|svgz|svi|swf|tar|tbz|tbz2|tcl|tga|tgz|thmx|tif|tiff|tk|tlz|topojson|torrent|ttc|ttf|txt|txz|udf|uvh|uvi|uvm|uvp|uvs|uvu|vcard|vcf|viv|vob|vtt|war|wav|wax|wbmp|wdp|weba|webapp|webm|webmanifest|webp|whl|wim|wm|wma|wml|wmlc|wmv|wmx|woff|woff2|wvx|xbm|xif|xla|xlam|xloc|xls|xlsb|xlsm|xlsx|xlt|xltm|xltx|xm|xmind|xml|xpi|xpm|xsl|xwd|xz|yuv|z|zip|zipx)(?:\?[a-zA-Z0-9\-\.\_\~\:\/\#\[\]\@\!\$\&\'\(\)\*\+\,\;\=]*)?$/i
};

var _warn = function warn (...args) {
  console.warn.call(console, '[WARN] [Spiderable-Middleware]', ...args);
};

request.defaultOptions.debug       = false;
request.defaultOptions.headers     = { 'User-Agent': 'spiderable-middleware/1.6.2', Accept: '*/*' };
request.defaultOptions.noStorage   = true;
request.defaultOptions.rawBody     = true;
request.defaultOptions.retry       = true;
request.defaultOptions.retries     = 3;
request.defaultOptions.retryDelay  = 128;
request.defaultOptions.timeout     = 102400;
request.defaultOptions.wait        = true;
request.defaultOptions.badStatuses = [502, 503, 504, 599];
request.defaultOptions.isBadStatus = function (statusCode, badStatuses) {
  return badStatuses.includes(statusCode);
};

module.exports = (function () {
  function Spiderable(_opts) {
    var opts = {};
    if (_opts && Object.prototype.toString.call(_opts) === strs.objs.object) {
      opts = _opts;
    }

    this.auth       = opts.auth;
    this.debug      = opts.debug || process.env.DEBUG === 'true' || process.env.DEBUG === true || false;
    var ignore      = opts.ignore || false;
    this.only       = opts.only || false;
    this.onlyRE     = opts.onlyRE || false;
    this.botsUA     = opts.botsUA || Spiderable.prototype.botsUA;
    this.rootURL    = opts.rootURL || process.env.ROOT_URL;
    this.timeout    = opts.timeout || 180000;
    this.staticExt  = opts.staticExt || re.staticExt;
    this.serviceURL = opts.serviceURL || process.env.SPIDERABLE_SERVICE_URL || process.env.PRERENDER_SERVICE_URL || 'https://render.ostr.io';
    this.ignoredHeaders = opts.ignoredHeaders || Spiderable.prototype.ignoredHeaders;
    this.requestOptions = opts.requestOptions || {};

    if (Object.prototype.toString.call(this.staticExt) !== strs.objs.regexp) {
      _warn('`opts.staticExt` must be instance of RegExp, falling back to defaults.');
      this.staticExt = re.staticExt;
    }

    if (this.onlyRE && Object.prototype.toString.call(this.onlyRE) !== strs.objs.regexp) {
      _warn('`opts.onlyRE` must be instance of RegExp, rules are ignored!');
      this.onlyRE = false;
    }

    if (Object.prototype.toString.call(this.botsUA) !== strs.objs.array) {
      _warn('`opts.botsUA` must be instance of Array, falling back to defaults.');
      this.botsUA = this.prototype.botsUA;
    }

    if (Object.prototype.toString.call(this.ignoredHeaders) !== strs.objs.array) {
      _warn('`opts.ignoredHeaders` must be instance of Array, falling back to defaults.');
      this.ignoredHeaders = this.prototype.ignoredHeaders;
    }

    if (this.only && Object.prototype.toString.call(this.only) !== strs.objs.array) {
      _warn('`opts.only` must be instance of Array, rules are ignored!');
      this.only = false;
    }

    if (!this.handler) {
      this.handler = this.middleware;
    }

    if (!this.handle) {
      this.handle = this.middleware;
    }

    this.headersRE = new RegExp('^(' + this.ignoredHeaders.join(strs.pipe) + ')$', strs.caseInsensitive);
    this.botsRE = new RegExp(this.botsUA.join(strs.pipe), strs.caseInsensitive);
    if (!this.auth) {
      this.auth = process.env.SPIDERABLE_SERVICE_AUTH || process.env.PRERENDER_SERVICE_AUTH || strs.empty;
    }

    if (ignore && Object.prototype.toString.call(ignore) !== strs.objs.array) {
      _warn('`opts.ignore` must be instance of Array, rules are ignored!');
      ignore = false;
    }

    if (!this.rootURL) {
      throw new Error('{rootURL} or env variable ROOT_URL is not detected! But must be specified!');
    }

    if (!this.serviceURL) {
      throw new Error('{serviceURL} or env variable SPIDERABLE_SERVICE_URL or PRERENDER_SERVICE_URL is not detected! But must be specified!');
    }

    if (!re.proto.test(this.rootURL)) {
      throw new Error('{rootURL} is malformed! Must start with protocol http or https');
    }

    if (!re.proto.test(this.serviceURL)) {
      throw new Error('{serviceURL} is malformed! Must start with protocol http or https');
    }

    this.rootURL    = this.rootURL.replace(re.trailingSlash, strs.empty).replace(re.beginningSlash, strs.empty);
    this.serviceURL = this.serviceURL.replace(re.trailingSlash, strs.empty).replace(re.beginningSlash, strs.empty);

    if (ignore) {
      this.ignoreRE = new RegExp(ignore.join(strs.pipe), strs.empty);
    } else {
      this.ignoreRE = false;
    }
  }

  Spiderable.prototype.botsUA = ['\\.net crawler', '360spider', '50\\.nu', '8bo crawler bot', 'aboundex', 'accoona', 'adldxbot', 'adsbot-google', 'ahrefsbot', 'altavista', 'appengine-google', 'applebot', 'archiver', 'arielisbot', 'ask jeeves', 'auskunftbot', 'baidumobaider', 'baiduspider', 'becomebot', 'bingbot', 'bingpreview', 'bitbot', 'bitlybot', 'blitzbot', 'blogbridge', 'boardreader', 'botseer', 'catchbot', 'catchpoint bot', 'charlotte', 'checklinks', 'cliqzbot', 'clumboot', 'coccocbot', 'converacrawler', 'crawl-e', 'crawlconvera', 'dataparksearch', 'daum', 'deusu', 'developers\\.google\\.com/+/web/snippet', 'discordbot', 'dotbot', 'duckduckbot', 'elefent', 'embedly', 'evernote', 'exabot', 'facebookbot', 'facebookexternalhit', 'fatbot', 'fdse robot', 'feed seeker bot', 'feedfetcher', 'femtosearchbot', 'findlinks', 'flamingo_searchengine', 'flipboard', 'followsite bot', 'furlbot', 'fyberspider', 'gaisbot', 'galaxybot', 'geniebot', 'genieo', 'gigablast', 'gigabot', 'girafabot', 'gomezagent', 'gonzo1', 'google sketchup', 'google-structured-data-testing-tool', 'googlebot', 'haosouspider', 'heritrix', 'holmes', 'hoowwwer', 'htdig', 'ia_archiver', 'idbot', 'infuzapp', 'innovazion crawler', 'instagram', 'internetarchive', 'iqdb', 'iskanie', 'istellabot', 'izsearch\\.com', 'kaloogabot', 'kaz\\.kz_bot', 'kd bot', 'konqueror', 'kraken', 'kurzor', 'larbin', 'leia', 'lesnikbot', 'linguee bot', 'linkaider', 'linkapediabot', 'linkedinbot', 'lite bot', 'llaut', 'lookseek', 'lycos', 'mail\\.ru_bot', 'masidani_bot', 'masscan', 'mediapartners-google', 'metajobbot', 'mj12bot', 'mnogosearch', 'mogimogi', 'mojeekbot', 'motominerbot', 'mozdex', 'msiecrawler', 'msnbot', 'msrbot', 'netpursual', 'netresearch', 'netvibes', 'newsgator', 'ng-search', 'nicebot', 'nutchcvs', 'nuzzel', 'nymesis', 'objectssearch', 'odklbot', 'omgili', 'oovoo', 'oozbot', 'openfosbot', 'orangebot', 'orbiter', 'org_bot', 'outbrain', 'pagepeeker', 'pagesinventory', 'parsijoobot', 'paxleframework', 'peeplo screenshot bot', 'pinterest', 'plantynet_webrobot', 'plukkie', 'pompos', 'psbot', 'quora link preview', 'qwantify', 'read%20later', 'reaper', 'redcarpet', 'redditbot', 'retreiver', 'riddler', 'rival iq', 'rogerbot', 'saucenao', 'scooter', 'scrapy', 'scrubby', 'searchie', 'searchsight', 'seekbot', 'semanticdiscovery', 'seznambot', 'showyoubot', 'simplepie', 'simpy', 'sitelockspider', 'skypeuripreview', 'slack-imgproxy', 'slackbot', 'slurp', 'snappy', 'sogou', 'solofield', 'speedy spider', 'speedyspider', 'sputnikbot', 'stackrambler', 'teeraidbot', 'teoma', 'theusefulbot', 'thumbshots\\.ru', 'thumbshotsbot', 'tineye', 'toweya\\.com', 'toweyabot', 'tumblr', 'tweetedtimes', 'tweetmemebot', 'twitterbot', 'url2png', 'vagabondo', 'vebidoobot', 'viber', 'visionutils', 'vkshare', 'voilabot', 'vortex', 'votay bot', 'voyager', 'w3c_validator', 'wasalive\\.bot', 'web-sniffer', 'websquash\\.com', 'webthumb', 'whatsapp', 'whatweb', 'wire', 'wotbox', 'yacybot', 'yahoo', 'yandex', 'yeti', 'yisouspider', 'yodaobot', 'yooglifetchagent', 'yoozbot', 'yottaamonitor', 'yowedo', 'zao-crawler', 'zebot_www\\.ze\\.bz', 'zooshot', 'zyborg'];

  Spiderable.prototype.ignoredHeaders = ['age', 'alt-svc', 'cache-status', 'cf-connecting-ip', 'cf-ipcountry', 'cf-cache-status', 'cf-ray', 'cf-request-id', 'cnection', 'cneonction', 'connection', 'content-encoding', 'content-length', 'date', 'etag', 'expect-ct', 'expires', 'keep-alive', 'last-modified', 'link', 'nel', 'nncoection', 'pragma', 'server', 'set-cookie', 'status', 'transfer-encoding', 'report-to', 'vary', 'via', 'www-authenticate', 'x-accel-buffering', 'x-accel-charset', 'x-accel-expires', 'x-accel-limit-rate', 'x-accel-redirect', 'x-ostrio-domain', 'x-powered-by', 'x-preprender-status', 'x-prerender-status', 'x-real-ip', 'x-runtime'];

  Spiderable.prototype.middleware = function (req, res, next) {
    var method = req.method.toLowerCase();
    if (method !== strs.get && method !== strs.head) {
      return next();
    }

    var urlObj = url.parse(req.url, true);
    if ((urlObj.query && urlObj.query._escaped_fragment_ !== void 0) || this.botsRE.test(req.headers[strs.ua] || strs.empty)) {
      var hasIgnored = false;
      var hasOnly    = false;

      if (this.staticExt.test(urlObj.pathname)) {
        return next();
      }

      if (this.onlyRE) {
        hasOnly    = this.onlyRE.test(urlObj.pathname);
        hasIgnored = !hasOnly;
      }

      if (!hasOnly && this.only) {
        hasIgnored = true;

        for (var i = 0; i < this.only.length; i++) {
          if (Object.prototype.toString.call(this.only[i]) === strs.objs.string) {
            if (this.only[i] === urlObj.pathname) {
              hasIgnored = false;
              hasOnly    = true;
              break;
            }
          } else if (Object.prototype.toString.call(this.only[i]) === strs.objs.regexp) {
            if (this.only[i].test(urlObj.pathname)) {
              hasIgnored = false;
              hasOnly    = true;
              break;
            }
          } else {
            _warn('`opts.only` {' + this.only[i] + '} rule isn\'t instance of {String} nor {RegExp}, rule ignored!');
          }
        }
      }

      if (this.ignoreRE && this.ignoreRE.test(urlObj.pathname)) {
        hasIgnored = true;
      }

      if (hasIgnored) {
        return next();
      }

      var reqUrl = this.rootURL;

      urlObj.path = urlObj.path.replace(re.trailingSlash, strs.empty).replace(re.beginningSlash, strs.empty);
      if (urlObj.query && typeof urlObj.query._escaped_fragment_ === 'string' && urlObj.query._escaped_fragment_.length) {
        urlObj.pathname += '/' + urlObj.query._escaped_fragment_.replace(re.beginningSlash, strs.empty);
      }

      reqUrl += '/' + urlObj.pathname;
      reqUrl  = reqUrl.replace(/([^:]\/)\/+/g, '$1');
      reqUrl  = (this.serviceURL + '/?url=' + encodeURIComponent(reqUrl));

      if (req.headers[strs.ua]) {
        reqUrl += '&bot=' + encodeURIComponent(req.headers[strs.ua]);
      }

      var opts  = Object.assign({}, this.requestOptions, {
        uri: reqUrl,
        auth: this.auth || false,
        debug: this.debug
      });

      try {
        var usedHeaders = [];
        var _headersRE  = this.headersRE;
        var serviceReq  = request(opts, function (error, resp) {
          if (error) {
            // DO NOT THROW AN ERROR ABOUT ABORTED REQUESTS
            if (!req.aborted) {
              _warn('Error while connecting to external service:', error);
              next();
            }
          } else {
            if (resp.statusCode === 401 || resp.statusCode === 403) {
              _warn('Can\'t authenticate! Please check you "auth" parameter and other settings.');
            }

            if (method === strs.head) {
              res.end();
            }
          }
        });

        serviceReq.onHeader(function (header) {
          var h = header.toString('utf8');
          if (h.includes(strs.semicolon)) {
            h = h.split(strs.semicolon);
            h[0] = h[0].trim().toLowerCase();
            h[1] = h[1].replace(re.newLine, strs.empty).trim();

            if (!res.headersSent && h[1].length && !usedHeaders.includes(h[0])) {
              if (h[0] === 'status') {
                var status = h[1].match(re.digit);
                if (status && status[0]) {
                  res.statusCode = status[0];
                  usedHeaders.push(h[0]);
                }
              } else if (!_headersRE.test(h[0])) {
                try {
                  res.setHeader(h[0], h[1]);
                } catch (e) {
                  _warn('.setHeader() Error:', e);
                }
                usedHeaders.push(h[0]);
              }
            }
          }
        });

        if (method !== strs.head) {
          serviceReq.pipe(res);
        }

        req.on('error', function (error) {
          _warn('[REQ] ["error" event] Unexpected error:', error);
          serviceReq.abort();
          next();
        });

        res.on('error', function (error) {
          _warn('[RES] ["error" event] Unexpected error:', error);
          serviceReq.abort();
          next();
        });

        req.on('aborted', function () {
          // No need to log this event as nothing bad happened
          // this simply means host which sent this request
          // has aborted the connection or got disconnected
          // _warn('[REQ] ["aborted" event]:', arguments); // TODO: comment out
          req.aborted = true;
          serviceReq.abort();
          try {
            res.end();
          } catch (e) {
            // We assume res.end() to throw an error
            // but still will call for it, as we'd like
            // to make sure memory, socket and session
            // are freed-up and closed
          }
        });

        // SET TIMEOUT AS A SECURITY PRECAUTION
        res.setTimeout(this.timeout);

        // res.on('close', function () {
        //   _warn('[RES] ["close" event]:', arguments);
        //   serviceReq.abort();
        //   next();
        // });

        serviceReq.send();
      } catch (e) {
        _warn('Exception while connecting to external service:', e);
        next();
      }
      return true;
    }
    return next();
  };

  return Spiderable;
})();
